File: pthreadfs/pthreadfs.cpp
#include "pthreadfs.h"

#include <assert.h>
#include <emscripten.h>
#include <pthread.h>
#include <wasi/api.h>

#include <thread>
#include <functional>
#include <iostream>

#include <stdarg.h>

SyncToAsync::SyncToAsync() : thread(threadMain, this), childLock(mutex) {
  // The child lock is associated with the mutex, which takes the lock, and
  // we free it here. Only the child will lock/unlock it from now on.
  childLock.unlock();
}

SyncToAsync::~SyncToAsync() {
  quit = true;

  shutdown();

  thread.join();
}

void SyncToAsync::shutdown() {
  readyToWork = true;
  condition.notify_one();
}

void SyncToAsync::doWork(std::function<void(SyncToAsync::Callback)> newWork) {
  // Send the work over.
  {
    std::lock_guard<std::mutex> lock(mutex);
    work = newWork;
    finishedWork = false;
    readyToWork = true;
  }
  condition.notify_one();

  // Wait for it to be complete.
  std::unique_lock<std::mutex> lock(mutex);
  condition.wait(lock, [&]() {
    return finishedWork;
  });
}

void* SyncToAsync::threadMain(void* arg) {
  // Prevent the pthread from shutting down too early.
  EM_ASM(runtimeKeepalivePush(););
  auto* parent = (SyncToAsync*)arg;
  emscripten_async_call(threadIter, arg, 0);
  return 0;
}

void SyncToAsync::threadIter(void* arg) {
  auto* parent = (SyncToAsync*)arg;
  // Wait until we get something to do.
  parent->childLock.lock();
  parent->condition.wait(parent->childLock, [&]() {
    return parent->readyToWork;
  });
  if (parent->quit) {
    EM_ASM(runtimeKeepalivePop(););
    return;
  }
  auto work = parent->work;
  parent->readyToWork = false;
  // Do the work.
  work([parent, arg]() {
    // We are called, so the work was finished. Notify the caller.
    parent->finishedWork = true;
    parent->childLock.unlock();
    parent->condition.notify_one();
    threadIter(arg);
  });
}

// Define global variables to be populated by resume;
SyncToAsync::Callback g_resumeFct;
SyncToAsync g_synctoasync_helper;

// Static functions calling resumFct and setting the return value.
void resumeWrapper_v()
{
  g_resumeFct();
}
// return value long
long resume_result_long = 0;
void resumeWrapper_l(long retVal)
{
  resume_result_long = retVal;
  g_resumeFct();
}
// return value __wasi_errno_t
__wasi_errno_t resume_result_wasi = 0;
void resumeWrapper_wasi(__wasi_errno_t retVal)
{
  resume_result_wasi = retVal;
  g_resumeFct();
}

// Wasi definitions
WASI_CAPI_DEF(write, const __wasi_ciovec_t *iovs, size_t iovs_len, __wasi_size_t *nwritten) {
  WASI_SYNCTOASYNC(write, iovs, iovs_len, nwritten);
}
WASI_CAPI_DEF(read, const __wasi_iovec_t *iovs, size_t iovs_len, __wasi_size_t *nread) {
  WASI_SYNCTOASYNC(read, iovs, iovs_len, nread);
}
WASI_CAPI_DEF(pwrite, const __wasi_ciovec_t *iovs, size_t iovs_len, __wasi_filesize_t offset, __wasi_size_t *nwritten) {
  WASI_SYNCTOASYNC(pwrite, iovs, iovs_len, offset, nwritten);
}
WASI_CAPI_DEF(pread, const __wasi_iovec_t *iovs, size_t iovs_len, __wasi_filesize_t offset, __wasi_size_t *nread) {
  WASI_SYNCTOASYNC(pread, iovs, iovs_len, offset, nread);
}
WASI_CAPI_DEF(seek, __wasi_filedelta_t offset, __wasi_whence_t whence, __wasi_filesize_t *newoffset) {
  WASI_SYNCTOASYNC(seek, offset, whence, newoffset);
}
WASI_CAPI_DEF(fdstat_get, __wasi_fdstat_t *stat) {
  WASI_SYNCTOASYNC(fdstat_get, stat);
}
WASI_CAPI_NOARGS_DEF(close) {
  WASI_SYNCTOASYNC_NOARGS(close);
}
WASI_CAPI_NOARGS_DEF(sync) {
  WASI_SYNCTOASYNC_NOARGS(sync);
}

// Syscall definitions
SYS_CAPI_DEF(open, 5, long path, long flags, ...) {
  va_list vl;
  va_start(vl, flags);
  int varargs = va_arg(vl, int);
  va_end(vl);
  
  SYS_SYNCTOASYNC(open, path, flags, varargs);
}

SYS_CAPI_DEF(unlink, 10, long path) {
  SYS_SYNCTOASYNC(unlink, path);
}

SYS_CAPI_DEF(chdir, 12, long path) {
  SYS_SYNCTOASYNC(chdir, path);
}

SYS_CAPI_DEF(mknod, 14, long path, long mode, long dev) {
  SYS_SYNCTOASYNC(mknod, path, mode, dev);
}

SYS_CAPI_DEF(chmod, 15, long path, long mode) {
  SYS_SYNCTOASYNC(chmod, path, mode);
}

SYS_CAPI_DEF(access, 33, long path, long amode) {
  SYS_SYNCTOASYNC(access, path, amode);
}

SYS_CAPI_DEF(rename, 38, long old_path, long new_path) {
  SYS_SYNCTOASYNC(rename, old_path, new_path);
}

SYS_CAPI_DEF(mkdir, 39, long path, long mode) {
  SYS_SYNCTOASYNC(mkdir, path, mode);
}

SYS_CAPI_DEF(rmdir, 40, long path) {
  SYS_SYNCTOASYNC(rmdir, path);
}

SYS_CAPI_DEF(dup, 41, long fd) {
  SYS_SYNCTOASYNC(dup, fd);
}

SYS_CAPI_DEF(ioctl, 54, long fd, long request, ...) {
  void *arg;
	va_list ap;
	va_start(ap, request);
	arg = va_arg(ap, void *);
	va_end(ap);
  
  SYS_SYNCTOASYNC(ioctl, fd, request, arg);
}

SYS_CAPI_DEF(dup2, 63, long oldfd, long newfd) {
  SYS_SYNCTOASYNC(dup2, oldfd, newfd);
}

SYS_CAPI_DEF(symlink, 83, long target, long linkpath) {
  SYS_SYNCTOASYNC(symlink, target, linkpath);
}

SYS_CAPI_DEF(readlink, 85, long path, long buf, long bufsize) {
  SYS_SYNCTOASYNC(readlink, path, buf, bufsize);
}

SYS_CAPI_DEF(munmap, 91, long addr, long len) {
  SYS_SYNCTOASYNC(munmap, addr, len);
}

SYS_CAPI_DEF(fchmod, 94, long fd, long mode) {
  SYS_SYNCTOASYNC(fchmod, fd, mode);
}

SYS_CAPI_DEF(fchdir, 133, long fd) {
  SYS_SYNCTOASYNC(fchdir, fd);
}

SYS_CAPI_DEF(_newselect, 142, long nfds, long readfds, long writefds, long exceptfds, long timeout) {
  SYS_SYNCTOASYNC(_newselect, nfds, readfds, writefds, exceptfds, timeout);
}

SYS_CAPI_DEF(msync, 144, long addr, long len, long flags) {
  SYS_SYNCTOASYNC(msync, addr, len, flags);
}

SYS_CAPI_DEF(fdatasync, 148, long fd) {
  SYS_SYNCTOASYNC(fdatasync, fd);
}

SYS_CAPI_DEF(poll, 168, long fds, long nfds, long timeout) {
  SYS_SYNCTOASYNC(poll, fds, nfds, timeout);
}

SYS_CAPI_DEF(getcwd, 183, long buf, long size) {
  SYS_SYNCTOASYNC(getcwd, buf, size);
}

SYS_CAPI_DEF(mmap2, 192, long addr, long len, long prot, long flags, long fd, long off) {
  SYS_SYNCTOASYNC(mmap2, addr, len, prot, flags, fd, off);
}

SYS_CAPI_DEF(truncate64, 193, long path, long zero, long low, long high) {
  SYS_SYNCTOASYNC(truncate64, path, zero, low, high);
}

SYS_CAPI_DEF(ftruncate64, 194, long fd, long zero, long low, long high) {
  SYS_SYNCTOASYNC(ftruncate64, fd, zero, low, high);
}

SYS_CAPI_DEF(stat64, 195, long path, long buf) {
  SYS_SYNCTOASYNC(stat64, path, buf);
}

SYS_CAPI_DEF(lstat64, 196, long path, long buf) {
  SYS_SYNCTOASYNC(lstat64, path, buf);
}

SYS_CAPI_DEF(fstat64, 197, long fd, long buf) {
  SYS_SYNCTOASYNC(fstat64, fd, buf);
}

SYS_CAPI_DEF(lchown32, 198, long path, long owner, long group) {
  SYS_SYNCTOASYNC(lchown32, path, owner, group);
}

SYS_CAPI_DEF(fchown32, 207, long fd, long owner, long group) {
  SYS_SYNCTOASYNC(fchown32, fd, owner, group);
}

SYS_CAPI_DEF(chown32, 212, long path, long owner, long group) {
  SYS_SYNCTOASYNC(chown32, path, owner, group);
}

SYS_CAPI_DEF(getdents64, 220, long fd, long dirp, long count) {
  SYS_SYNCTOASYNC(getdents64, fd, dirp, count);
}

SYS_CAPI_DEF(fcntl64, 221, long fd, long cmd, ...) {
va_list vl;
  va_start(vl, cmd);
  int varargs = va_arg(vl, int);
  va_end(vl);
  
  SYS_SYNCTOASYNC(fcntl64, fd, cmd, varargs);
}

SYS_CAPI_DEF(statfs64, 268, long path, long size, long buf) {
  SYS_SYNCTOASYNC(statfs64, path, size, buf);
}

SYS_CAPI_DEF(fstatfs64, 269, long fd, long size, long buf) {
  SYS_SYNCTOASYNC(fstatfs64, fd, size, buf);
}

SYS_CAPI_DEF(openat, 295, long dirfd, long path, long flags, ...) {
va_list vl;
  va_start(vl, flags);
  int varargs = va_arg(vl, int);
  va_end(vl);
  
  SYS_SYNCTOASYNC(openat, dirfd, path, flags, varargs);
}

SYS_CAPI_DEF(mkdirat, 296, long dirfd, long path, long mode) {
  SYS_SYNCTOASYNC(mkdirat, dirfd, path, mode);
}

SYS_CAPI_DEF(mknodat, 297, long dirfd, long path, long mode, long dev) {
  SYS_SYNCTOASYNC(mknodat, dirfd, path, mode, dev);
}

SYS_CAPI_DEF(fchownat, 298, long dirfd, long path, long owner, long group, long flags) {
  SYS_SYNCTOASYNC(fchownat, dirfd, path, owner, group, flags);
}

SYS_CAPI_DEF(fstatat64, 300, long dirfd, long path, long buf, long flags) {
  SYS_SYNCTOASYNC(fstatat64, dirfd, path, buf, flags);
}

SYS_CAPI_DEF(unlinkat, 301, long dirfd, long path, long flags) {
  SYS_SYNCTOASYNC(unlinkat, dirfd, path, flags);
}

SYS_CAPI_DEF(renameat, 302, long olddirfd, long oldpath, long newdirfd, long newpath) {
  SYS_SYNCTOASYNC(renameat, olddirfd, oldpath, newdirfd, newpath);
}

SYS_CAPI_DEF(symlinkat, 304, long target, long newdirfd, long linkpath) {
  SYS_SYNCTOASYNC(symlinkat, target, newdirfd, linkpath);
}

SYS_CAPI_DEF(readlinkat, 305, long dirfd, long path, long bug, long bufsize) {
  SYS_SYNCTOASYNC(readlinkat, dirfd, path, bug, bufsize);
}

SYS_CAPI_DEF(fchmodat, 306, long dirfd, long path, long mode, ...) {
va_list vl;
  va_start(vl, mode);
  int varargs = va_arg(vl, int);
  va_end(vl);
  
  SYS_SYNCTOASYNC(fchmodat, dirfd, path, mode, varargs);
}

SYS_CAPI_DEF(faccessat, 307, long dirfd, long path, long amode, long flags) {
  SYS_SYNCTOASYNC(faccessat, dirfd, path, amode, flags);
}

SYS_CAPI_DEF(utimensat, 320, long dirfd, long path, long times, long flags) {
  SYS_SYNCTOASYNC(utimensat, dirfd, path, times, flags);
}

SYS_CAPI_DEF(fallocate, 324, long fd, long mode, long off_low, long off_high, long len_low, long len_high) {
  SYS_SYNCTOASYNC(fallocate, fd, mode, off_low, off_high, len_low, len_high);
}

SYS_CAPI_DEF(dup3, 330, long fd, long suggestfd, long flags) {
  SYS_SYNCTOASYNC(dup3, fd, suggestfd, flags);
}

// Other helper code

void emscripten_init_pthreadfs() {
  g_synctoasync_helper.doWork([](SyncToAsync::Callback resume) {
    g_resumeFct = [resume]() { resume(); };
    init_pthreadfs(&resumeWrapper_v);
  });
  g_synctoasync_helper.doWork([](SyncToAsync::Callback resume) {
    g_resumeFct = [resume]() { resume(); };
    init_fsafs(&resumeWrapper_v);
  });
  return;
}

File: pthreadfs/src/js/pthreadfs.js
/**
 * @license
 * Copyright 2021 The Emscripten Authors
 * SPDX-License-Identifier: MIT
 */

var SyscallWrappers = {}

let SyscallsFunctions = [
  {'name': 'open', 'args': ['path', 'flags', 'varargs']},
  {'name': 'unlink', 'args': ['path']},
  {'name': 'chdir', 'args': ['path']},
  {'name': 'mknod', 'args': ['path', 'mode', 'dev']},
  {'name': 'chmod', 'args': ['path', 'mode']},
  {'name': 'access', 'args': ['path', 'amode']},
  {'name': 'rename', 'args': ['old_path', 'new_path']},
  {'name': 'mkdir', 'args': ['path', 'mode']},
  {'name': 'rmdir', 'args': ['path']},
  {'name': 'dup', 'args': ['fd']},
  {'name': 'ioctl', 'args': ['fd', 'request', 'varargs']},
  {'name': 'dup2', 'args': ['oldfd', 'newfd']},
  {'name': 'symlink', 'args': ['target', 'linkpath']},
  {'name': 'readlink', 'args': ['path', 'buf', 'bufsize']},
  {'name': 'munmap', 'args': ['addr', 'len']},
  {'name': 'fchmod', 'args': ['fd', 'mode']},
  {'name': 'fchdir', 'args': ['fd']},
  {'name': '_newselect', 'args': ['nfds', 'readfds', 'writefds', 'exceptfds', 'timeout']},
  {'name': 'msync', 'args': ['addr', 'len', 'flags']},
  {'name': 'fdatasync', 'args': ['fd']},
  {'name': 'poll', 'args': ['fds', 'nfds', 'timeout']},
  {'name': 'getcwd', 'args': ['buf', 'size']},
  {'name': 'mmap2', 'args': ['addr', 'len', 'prot', 'flags', 'fd', 'off']},
  {'name': 'truncate64', 'args': ['path', 'zero', 'low', 'high']},
  {'name': 'ftruncate64', 'args': ['fd', 'zero', 'low', 'high']},
  {'name': 'stat64', 'args': ['path', 'buf']},
  {'name': 'lstat64', 'args': ['path', 'buf']},
  {'name': 'fstat64', 'args': ['fd', 'buf']},
  {'name': 'lchown32', 'args': ['path', 'owner', 'group']},
  {'name': 'fchown32', 'args': ['fd', 'owner', 'group']},
  {'name': 'chown32', 'args': ['path', 'owner', 'group']},
  {'name': 'getdents64', 'args': ['fd', 'dirp', 'count']},
  {'name': 'fcntl64', 'args': ['fd', 'cmd', 'varargs']},
  {'name': 'statfs64', 'args': ['path', 'size', 'buf']},
  {'name': 'fstatfs64', 'args': ['fd', 'size', 'buf']},
  {'name': 'openat', 'args': ['dirfd', 'path', 'flags', 'varargs']},
  {'name': 'mkdirat', 'args': ['dirfd', 'path', 'mode']},
  {'name': 'mknodat', 'args': ['dirfd', 'path', 'mode', 'dev']},
  {'name': 'fchownat', 'args': ['dirfd', 'path', 'owner', 'group', 'flags']},
  {'name': 'fstatat64', 'args': ['dirfd', 'path', 'buf', 'flags']},
  {'name': 'unlinkat', 'args': ['dirfd', 'path', 'flags']},
  {'name': 'renameat', 'args': ['olddirfd', 'oldpath', 'newdirfd', 'newpath']},
  {'name': 'symlinkat', 'args': ['target', 'newdirfd', 'linkpath']},
  {'name': 'readlinkat', 'args': ['dirfd', 'path', 'bug', 'bufsize']},
  {'name': 'fchmodat', 'args': ['dirfd', 'path', 'mode', 'varargs']},
  {'name': 'faccessat', 'args': ['dirfd', 'path', 'amode', 'flags']},
  {'name': 'utimensat', 'args': ['dirfd', 'path', 'times', 'flags']},
  {'name': 'fallocate', 'args': ['fd', 'mode', 'off_low', 'off_high', 'len_low', 'len_high']},
  {'name': 'dup3', 'args': ['fd', 'suggestfd', 'flags']},
]

let WasiFunctions = [
  {'name': 'write', 'args': ['iovs', 'iovs_len', 'nwritten']},
  {'name': 'read', 'args': ['iovs', 'iovs_len', 'nread']},
  {'name': 'close', 'args': []},
  {'name': 'pwrite', 'args': ['iov', 'iovcnt', "{{{ defineI64Param('offset') }}}", 'pnum']},
  {'name': 'pread', 'args': ['iov', 'iovcnt', "{{{ defineI64Param('offset') }}}", 'pnum']},
  {'name': 'seek', 'args': ["{{{ defineI64Param('offset') }}}", 'whence', 'newOffset']},
  {'name': 'fdstat_get', 'args': ['pbuf']},
  {'name': 'sync', 'args': []},
]

function createWasiWrapper(name, args, wrappers) {
  let full_args = 'fd';
  if (args.length > 0) {
    full_args = full_args + ',' + args.join(',');
  }
  let full_args_with_resume = full_args + ',resume';
  let wrapper = `function(${full_args_with_resume}) {`;
  wrapper += `_fd_${name}_async(${full_args}).then((res) => {`;
  wrapper += 'wasmTable.get(resume)(res);});}'
  wrappers[`__fd_${name}_async`] = eval('(' + wrapper + ')');
  wrappers[`__fd_${name}_async__deps`] = [`fd_${name}_async`, '$ASYNCSYSCALLS', '$FSAFS'];
}

function createSyscallWrapper(name, args, wrappers) {
  let full_args = '';
  let full_args_with_resume = 'resume';
  if (args.length > 0) {
    full_args = args.join(',');
    full_args_with_resume = full_args + ', resume';
  }
  let wrapper = `function(${full_args_with_resume}) {`;
  wrapper += `_${name}_async(${full_args}).then((res) => {`;
  wrapper += 'wasmTable.get(resume)(res);});}'
  wrappers[`__sys_${name}_async`] = eval('(' + wrapper + ')');
  wrappers[`__sys_${name}_async__deps`] = [`${name}_async`, '$ASYNCSYSCALLS', '$FSAFS'];
}

for (x of WasiFunctions) {
  createWasiWrapper(x.name, x.args, SyscallWrappers);
}
for (x of SyscallsFunctions) {
  createSyscallWrapper(x.name, x.args, SyscallWrappers);
}

SyscallWrappers['init_pthreadfs'] = function (resume) {
  var FSNode = /** @constructor */ function(parent, name, mode, rdev) {
    if (!parent) {
      parent = this;  // root node sets parent to itself
    }
    this.parent = parent;
    this.mount = parent.mount;
    this.mounted = null;
    this.id = PThreadFS.nextInode++;
    this.name = name;
    this.mode = mode;
    this.node_ops = {};
    this.stream_ops = {};
    this.rdev = rdev;
  };
  var readMode = 292/*{{{ cDefine("S_IRUGO") }}}*/ | 73/*{{{ cDefine("S_IXUGO") }}}*/;
  var writeMode = 146/*{{{ cDefine("S_IWUGO") }}}*/;
  Object.defineProperties(FSNode.prototype, {
   read: {
    get: /** @this{FSNode} */function() {
     return (this.mode & readMode) === readMode;
    },
    set: /** @this{FSNode} */function(val) {
     val ? this.mode |= readMode : this.mode &= ~readMode;
    }
   },
   write: {
    get: /** @this{FSNode} */function() {
     return (this.mode & writeMode) === writeMode;
    },
    set: /** @this{FSNode} */function(val) {
     val ? this.mode |= writeMode : this.mode &= ~writeMode;
    }
   },
   isFolder: {
    get: /** @this{FSNode} */function() {
     return PThreadFS.isDir(this.mode);
    }
   },
   isDevice: {
    get: /** @this{FSNode} */function() {
     return PThreadFS.isChrdev(this.mode);
    }
   }
  });
  PThreadFS.FSNode = FSNode;

  PThreadFS.staticInit().then(async ()=> {
    await PThreadFS.init();
    PThreadFS.ignorePermissions = false;
    wasmTable.get(resume)();
  });
}

SyscallWrappers['init_fsafs'] = function(resume) {
  PThreadFS.mkdir('/filesystemaccess').then(async () => {
    await PThreadFS.mount(FSAFS, { root: '.' }, '/filesystemaccess');
    wasmTable.get(resume)();
  });
}

SyscallWrappers['init_sfafs'] = function(resume) {
  PThreadFS.mkdir('/sfa').then(async () => {
    await PThreadFS.mount(SFAFS, { root: '.' }, '/sfa');

    // Storage Foundation requires explicit capacity allocations.
    if (storageFoundation.requestCapacity) {
      await storageFoundation.requestCapacity(1024*1024*100);
    }
    // Delete all old files.
    let files = await storageFoundation.getAll();
    for (file of files) {
      await storageFoundation.delete(file);
    }
    wasmTable.get(resume)();
  });
}

mergeInto(LibraryManager.library, SyscallWrappers);